src/display/canvas.js

```diff
@@ -2120,17 +2122,19 @@ var CanvasGraphics = (function CanvasGraphicsClosure() {
           newHeight = Math.ceil(paintHeight / 2);
           heightScale /= paintHeight / newHeight;
         }
-        tmpCanvas = this.cachedCanvases.getCanvas(tmpCanvasId,
-                                                  newWidth, newHeight);
-        tmpCtx = tmpCanvas.context;
-        tmpCtx.clearRect(0, 0, newWidth, newHeight);
-        tmpCtx.drawImage(imgToPaint, 0, 0, paintWidth, paintHeight,
-                                     0, 0, newWidth, newHeight);
-        imgToPaint = tmpCanvas.canvas;
+
         paintWidth = newWidth;
         paintHeight = newHeight;
-        tmpCanvasId = tmpCanvasId === 'prescale1' ? 'prescale2' : 'prescale1';
       }
+
+      tmpCanvas = this.cachedCanvases.getCanvas(tmpCanvasId,
```

Comment:
The reason for the iterative downsampling is that most browsers do a terrible job of downsampling by factors larger than 2 (I guess, they do use nearest neighbors). PDF documents often contain (scanned) text and drawings, rather than fotos, so the browser's downsampling does in fact severely reduce image quality. 
See https://github.com/mozilla/pdf.js/issues/7041#issuecomment-192496865 and below for some examples in which I compare iterative downsampling vs. browser downsampling.